futures-core/src/task/__internal/atomic_waker.rs

```diff
@@ -214,35 +283,40 @@ impl AtomicWaker {
                         REGISTERING, WAITING, AcqRel, Acquire);
 
                     match res {
-                        Ok(_) => {}
+                        Ok(_) => {
+                            // memory ordering: acquired self.state
+                            // - if a previous wake went through it syncs with
+                            //   its final release (`fetch_and`)
+                            // - if there was no previous wake the next wake
+                            //   will wake us, no sync needed.
+                        }
                         Err(actual) => {
                             // This branch can only be reached if a
                             // concurrent thread called `wake`. In this
                             // case, `actual` **must** be `REGISTERING |
                             // `WAKING`.
                             debug_assert_eq!(actual, REGISTERING | WAKING);
 
-                            // Take the waker to wake once the atomic operation has
-                            // completed.
-                            let waker = (*self.waker.get()).take().unwrap();
+                            // Undo registration, no need to wake up later.
+                            let _waker = (*self.waker.get()).take().unwrap();
 
                             // Just swap, because no one could change state while state == `REGISTERING` | `WAKING`.
                             self.state.swap(WAITING, AcqRel);
 
-                            // The atomic swap was complete, now
-                            // wake the task and return.
-                            waker.wake();
+                            // memory ordering: we acquired self.state for a
+                            // wake() releasing self.state, all is fine.
                         }
                     }
                 }
             }
             WAKING => {
-                // Currently in the process of waking the task, i.e.,
-                // `wake` is currently being called on the old task handle.
-                // So, we call wake on the new waker
-                waker.wake_by_ref();
```

Comment:
With this change, if you called register the same time another thread was waking, the new waker would never get the wakeup.

Reply:
That is certainly true.  But: if the other thread was waking just a little bit earlier, the new waker wouldn't get the wakeup either - and this is fine, because it doesn't have to: we just acquired a state that was relased by the other thread "waking", so the memory is synced as promised.

Reply:
Damn. Those wakes are actually needed in case two waking threads ("producers") come along at the same time, and the first produced item has been consumed (without a call to `register`, which can happen). `register` will only sync with the first producer, and the second item might not be visible yet. I'll try to come up with some descriptions of those scenarious...

As far as I can see now:
- the first wake could probably be removed if the registration wasn't undone; all wakes during a registration should be properly synced, only the additional wakes after the registration are not synced (and don't find a registered waker right now).
- the second wake is probably not enough (strictly speaking), as Acquire/Release doesn't establish a transitive sync across threads, but another round through the event loop might simply take enough time.